File: C#/OOP.cs
using System;

//abstract class - may contain non abstract method.
abstract class Abstract{
    public Abstract(){
        Console.WriteLine("Abstract constructor is called.");
    }
    ~Abstract(){
        Console.WriteLine("Abstract destructor is called.");
    }
    public abstract string AbstractMethod();
    public string PredefinedMethod(){
        return "this is a predefined method from abstract class.";
    }
}

class Base: Abstract{
    private string name = "Base class method.";
    public Base(){
        Console.WriteLine("Base class constructor is called.");
    }
    ~Base(){
        Console.WriteLine("Base class destructor is called.");
    }
    // to mark as override-able function
    public virtual string StringMethod(){
        return this.name;
    }
    // override AbstractMethod
    public override string AbstractMethod(){
        return "AbstractMethod is called from class Abstract";
    }
}
// inheritance 
class Derived: Base{
    private string name = "Derived class method.";
    public Derived(){
        Console.WriteLine("Derived constructor is called.");
    }
    ~Derived(){
        Console.WriteLine("Derived class destructor is called.");
    }
    //compile-time polymorphism (function overloading)
    public Derived(string name){
        Console.WriteLine("Second variant of constructor is called containing " + name);
    }
    //runtime polymorphism (function overriding)
    public override string StringMethod(){
        return this.name;
    }
    //implementation of AbstractMethod
    public override string AbstractMethod(){
        return "AbstractMethod is called from class Derived";
    }
}

class MainClass{
    public static void Main(string [] args){
        Base b = new Base(); //instantiates base class
        Derived d = new Derived(); //instatiates derived class
        function(b);
        function(d);
    }
    public static void function(Base obj){ // you cannot parameterize an Abstract class
        Console.WriteLine(obj.StringMethod());
        Console.WriteLine(obj.AbstractMethod());
        Console.WriteLine(obj.PredefinedMethod());
    }
}

File: Python/OOP.py
from abc import ABC, abstractmethod


class Abstract(ABC):
    @abstractmethod
    def abstract_method(self):
        pass


class Base(Abstract):
    def __init__(self) -> None:
        print("Base class is called.")

    def method(self, a: int) -> None:
        print("prints {0} from Base class".format(a))

    def abstract_method(self):
        print("abstract_method is called from Base class.")

    def __del__(self) -> None:
        print("Base destructor is called")


class Derived(Base):
    def __init__(self) -> None:
        print("Derived constructor is called.")

    def method(self, a: int) -> None:
        print("prints {0} from Derived class".format(a))

    def abstract_method(self) -> None:
        print("abstract_method is called from Derived class")

    def __del__(self):
        print("Derived destructor is called")


def functon(obj: Abstract) -> None:
    return obj.abstract_method()


if __name__ == '__main__':
    base = Base()
    derived = Derived()

    base.method(3)
    derived.method(3)